src/gui_common/AddWindowReorderingSupportToSiblings.cs

```diff
@@ -0,0 +1,375 @@
+
+        connectedWindows.Add(window, topNode);
+        connectedSiblings.Add(topNode);
+
+        // Update top sibling
+        if (topSibling == null || topSibling.GetIndex() < topNode.GetIndex())
+            topSibling = topNode;
+    }
+
+    public void DisconnectWindow(CustomDialog window, bool recursive)
+    {
+        if (recursive)
+        {
+            // Connection was recursive, disconnect from other window reordering nodes
+            foreach (var ancestor in windowReorderingAncestors)
+            {
+                ancestor.ReorderingNode.DisconnectWindow(window, recursive);
+            }
+        }
+
+        if (!window.IsConnected(nameof(Dragged), this, nameof(OnWindowReorder)))
+        {
+            // This window has already been disconnected from here
+            return;
+        }
+
+        window.Disconnect(nameof(Dragged), this, nameof(OnWindowReorder));
+
+        var windowSibling = connectedWindows[window];
+        connectedWindows.Remove(window);
+
+        // Find if another connected window has the same sibling
+        bool foundSibling = false;
+        foreach (var sibling in connectedSiblings)
+        {
+            if (sibling == windowSibling)
+            {
+                foundSibling = true;
+                break;
+            }
+        }
+
+        if (!foundSibling)
+        {
+            // No other window has the same sibling so it can be removed
+            connectedSiblings.Remove(windowSibling);
+
+            if (topSibling == windowSibling)
+                topSibling = null;
+        }
+    }
+
+    /// <summary>
+    ///   Finds whose windows ancestor is currently on the top
+    /// </summary>
+    public void UpdateTopSibling()
+    {
+        if (topSibling != null)
+        {
+            // Top sibling is already current
+            return;
+        }
+
+        var childCount = parent.GetChildCount();
+
+        // Search for the first sibling that is part of the reordering system
+        for (int i = childCount - 1; i >= 0; --i)
+        {
+            var sibling = parent.GetChild(i);
+            if (!connectedSiblings.Contains(sibling))
+                continue;
+
+            topSibling = sibling;
+            break;
+        }
+    }
+
+    /// <summary>
+    ///   Reoders a window by setting its ancestor to the position of current top window, making it appear on top of
+    ///   others.
+    /// </summary>
+    public void OnWindowReorder(CustomDialog window)
+    {
+        // Get a sibling that is an ancestor of this window
+        Node targetSibling = connectedWindows[window];
+
+        UpdateTopSibling();
+
+        int topSiblingIndex = topSibling!.GetIndex();
+
+        if (topSiblingIndex == targetSibling.GetIndex())
+        {
+            // This window is already on the top
+            return;
+        }
+
+        // Put the sibling on the top
+        parent.MoveChild(targetSibling, topSiblingIndex);
```

Comment:
Can't we just simply make the index be 0, as that should always move any node to the top?

Reply:
I explicitly requested that this PR wouldn't reorder non-window nodes. Setting index to 0 might put some GUI stuff before nodes that want to be ahead of the GUI, like most of our stage scenes are setup like. So this is to ensure that only the window positions will jump to each other, and not ahead of other non-window nodes.

Reply:
I see, I somehow thought that all reorderable windows are within the scope of AddWindowReorderingSupportToSiblings (I see there's a dedicated node for it) just like how ModalManager works.

Reply:
Well not exactly. Especially as the following case needs to work:

- Some Control
  - ReorderSiblingsNode
  - Window
  - Window
  - Control
    - ReorderSiblingsNode
    - Window
    - Window
  - Window

So if the first Window inside the sub-control is clicked then the whole sub-control tree needs to move in its parent and also the two windows in it need to swap places.